modcc/module.cpp

```diff
@@ -699,114 +699,135 @@ void Module::add_variables_to_symbols() {
+                        std::cout << green("\n-argument lowering-\n\n");
+    #endif
+                    // lower function arguments that are not identifiers or literals
+                    // e.g.
+                    //      ll0_ = foo(2+x, y, 1)
+                    //      a = 2 + ll0_
+                    // becomes
+                    //      ll1_ = 2+x
+                    //      ll0_ = foo(ll1_, y, 1)
+                    //      a = 2 + ll0_
+                    for (auto e = b.begin(); e != b.end(); ++e) {
+                        if (auto be = (*e)->is_binary()) {
+                            // only apply to assignment expressions where rhs is a
+                            // function call because the function call lowering step
+                            // above ensures that all function calls are of this form
+                            if (auto rhs = be->rhs()->is_function_call()) {
+                                b.splice(e, lower_function_arguments(rhs->args()));
+                            }
                         }
                     }
+
+    #ifdef LOGGING
+                    std::cout << "body after argument lowering\n";
+                        for(auto& l : b) std::cout << "  " << l->to_string() << " @ " << l->location() << "\n";
+                        std::cout << green("\n-inlining-\n\n");
+    #endif
                 }
+            }
+        }
 
-#ifdef LOGGING
-                std::cout << "body after argument lowering\n";
-                for(auto& l : b) std::cout << "  " << l->to_string() << " @ " << l->location() << "\n";
-                std::cout << green("\n-inlining-\n\n");
-#endif
-
-                // Do the inlining, which currently only works for functions
-                // that have a single statement in their body
-                // e.g. if the function foo in the examples above is defined as follows
-                //
-                //  function foo(a, b, c) {
-                //      foo = a*(b + c)
-                //  }
-                //
-                // the full inlined example is
-                //      ll1_ = 2+x
-                //      ll0_ = ll1_*(y + 1)
-                //      a = 2 + ll0_
-                for(auto e=b.begin(); e!=b.end(); ++e) {
-                    if(auto ass = (*e)->is_assignment()) {
-                        if(ass->rhs()->is_function_call()) {
-                            ass->replace_rhs(inline_function_call(ass->rhs()));
+        for(auto& e : symbols_) {
+            auto& s = e.second;
+
+            if(s->kind() == symbolKind::procedure)
+            {
+                if(errors==0) {
+                    auto &b = s->kind()==symbolKind::function ?
+                        s->is_function()->body()->statements() :
+                        s->is_procedure()->body()->statements();
+
+                    // Do the inlining, which currently only works for functions
```

Comment:
I think that this comment is out of date after this PR.

Reply:
I updated it.